dlup/writers.py

```diff
@@ -0,0 +1,209 @@
+    LZW = "lzw"  # LZW compression, not implemented in tifffile
+    WEBP = "webp"  # WEBP compression
+    ZSTD = "zstd"  # ZSTD compression
+    JP2K = "jp2k"  # JP2K compression
+    JP2K_LOSSY = "jp2k_lossy"
+    PNG = "png"
+
+
+# Mapping to map TiffCompression to their respective values in tifffile.
+TIFFFILE_COMPRESSION = {
+    "none": None,
+    "ccittfax4": "CCITT_T4",
+    "jpeg": "jpeg",
+    "deflate": "deflate",
+    "packbits": "packbits",
+    "lzw": "lzw",
+    "webp": "webp",
+    "zstd": "zstd",
+    "jp2k": "jpeg2000",
+    "jp2k_lossy": "jpeg_2000_lossy",
+    "png": "png",
+}
+
+
+class ImageWriter:
+    """Base writer class"""
+
+
+class TifffileImageWriter(ImageWriter):
+    """Image writer that writes tile-by-tile to tiff."""
+
+    def __init__(
+        self,
+        filename: PathLike,
+        size: Union[Tuple[int, int, int], Tuple[int, int]],
+        mpp: Union[float, Tuple[float, float]],
+        tile_size: Tuple[int, int],
+        pyramid: bool = False,
+        compression: Optional[TiffCompression] = TiffCompression.JPEG,
+        quality: Optional[int] = 100,
+    ):
+        """
+        Writer based on tifffile.
+
+        Parameters
+        ----------
+        filename : PathLike
+            Filename where to write
+        size : Tuple
+            Size of the image to be written. This is defined as (height, width, num_channels),
+            or rather (rows, columns, num_channels) and is important value to get correct.
+            In case of a mask with a single channel the value is given by (rows, columns).
+        mpp : int, or Tuple[int, int]
+        tile_size : Tuple[int, int]
+            Tiff tile_size, defined as (height, width).
+        pyramid : bool
+            Whether to write a pyramidal image.
+        compression : TiffCompression
+            Compressor to use.
+        quality : int
+            Quality in case a lossy compressor is used.
+        """
+        self._filename = filename
+        self._tile_size = tile_size
+        self._size = (*size, 1) if len(size) == 2 else size
+
+        if isinstance(mpp, float):
+            mpp = (mpp, mpp)
+        self._mpp: Tuple[float, float] = mpp
+
+        if not compression:
+            compression = TiffCompression.NONE
+
+        self._compression = compression
+        self._pyramid = pyramid
+        self._quality = quality
+
+    def from_pil(self, pil_image: PIL.Image):
+        if not np.all(np.asarray(pil_image.size)[::-1] >= self._tile_size):
+            raise RuntimeError(
+                f"PIL Image must be larger than set tile size. Got {pil_image.size} and {self._tile_size}."
+            )
+        iterator = _pil_grid_iterator(pil_image, self._tile_size)
+        self.from_iterator(iterator, filename=self._filename)
+
+    def from_iterator(self, iterator: Iterator[np.ndarray | None]):
+        filename = pathlib.Path(self._filename)
+        temp_filename = filename.with_suffix(f"{filename.suffix}.partial")
```

Comment:
Is it intentional that you want to leave the .partial file, even if the program breaks? Otherwise I sould suggest go for a tempfile.